ospackage/recipes/artifacts.go

```diff
@@ -0,0 +1,182 @@
+// Returns a map of artifact names to their new locations on the local disk.
+func FetchArtifacts(ctx context.Context, artifacts []*osconfigpb.SoftwareRecipe_Artifact, directory string) (map[string]string, error) {
+	localNames := make(map[string]string)
+
+	for _, a := range artifacts {
+		path, err := fetchArtifact(ctx, a, directory)
+		if err != nil {
+			return nil, err
+		}
+		localNames[a.Id] = path
+	}
+
+	return localNames, nil
+}
+
+func fetchArtifact(ctx context.Context, artifact *osconfigpb.SoftwareRecipe_Artifact, directory string) (string, error) {
+	localPath := path.Join(directory, artifact.Id)
+	uri, err := url.Parse(artifact.Uri)
+	if err != nil {
+		return "", fmt.Errorf("Could not parse url %q for artifact %q", artifact.Uri, artifact.Id)
+	}
+
+	var reader io.ReadCloser
+	switch strings.ToLower(uri.Scheme) {
+	case "gcs":
+		reader, err = fetchWithGCS(ctx, uri.Host, uri.Path, uri.Fragment)
+		if err != nil {
+			return "", fmt.Errorf("error fetching artifact %q from GCS: %v", artifact.Id, err)
+		}
+		defer reader.Close()
+	case "https", "http":
+		response, err := fetchWithHTTP(ctx, artifact.Uri)
+		if err != nil {
+			return "", fmt.Errorf("error fetching artifact %q with http or https: %v", artifact.Id, err)
+		}
+		defer response.Body.Close()
+		reader = response.Body
+	default:
+		return "", fmt.Errorf("artifact %q has unsupported protocol %s", artifact.Id, uri.Scheme)
+	}
+
+	err = downloadStream(reader, artifact.Checksum, localPath)
+	if err != nil {
+		return "", err
+	}
+	return localPath, nil
+}
+
+func fetchWithGCS(ctx context.Context, bucket, path, generation string) (*storage.Reader, error) {
+	client, err := newStorageClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create storage client: %v", err)
+	}
+	defer client.Close()
+
+	oh := client.Bucket(bucket).Object(path)
+	if generation != "" {
+		generationNumber, err := strconv.ParseInt(generation, 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("couldn't parse gcs generation number %q", generation)
+		}
+		oh = oh.Generation(generationNumber)
+	}
+
+	r, err := oh.NewReader(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return r, nil
+}
+
+func fetchWithHTTP(ctx context.Context, uri string) (*http.Response, error) {
+	resp, err := newHTTPClient().Get(uri)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("got http status %d when attempting to download artifact", resp.StatusCode)
+	}
+
+	return resp, nil
+}
+
+func downloadStream(r io.Reader, checksum string, localPath string) error {
+	localPath, err := normPath(localPath)
+	if err != nil {
+		return err
+	}
+	file, err := os.Create(localPath)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	hasher := sha256.New()
+	_, err = io.Copy(io.MultiWriter(file, hasher), r)
+	if err != nil {
+		return err
+	}
+	computed := fmt.Sprintf("%64x", hasher.Sum(nil))
```

Comment:
Typically hex.EncodeToString is used, it doesnt require you to count if the hasher is changed

Reply:
changed